server/serverMethods.js

```diff
@@ -619,51 +628,69 @@ function ServerMethods(aLogLevel, aModules) {
+    var verifyIdTokenPromise = Utils.promisify(client.verifyIdToken, 1, client);
+
+    verifyIdTokenPromise(body.googleIdToken, tbConfig.googleId)
+      .then((login) => {
+        var payload = login.getPayload();
+        var domain = payload.hd;
+        if (domain !== tbConfig.googleHostedDomain) {
+          logger.log('postRoomDial => authenticated token domain did not match config: ', domain);
+          aRes.status(403).send(new ErrorInfo(100, 'Authenticaton Domain Invalid'));
+        } else {
+          var phoneNumber = body.phoneNumber;
+          if (dialedNumbersInfo[phoneNumber]) {
+            return;
+          }
+          var otInstance = tbConfig.otInstance;
+
+          serverPersistence
+            .getKey(redisRoomPrefix + roomName)
+            .then((sessionInfo) => {
+              const sessionId = JSON.parse(sessionInfo).sessionId;
+              const token = otInstance.generateToken(sessionId, {
+                role: 'publisher',
+                data: '{"sip":true, "role":"client", "name":"' + phoneNumber + '"}'
+              });
+              var options = {
+                // Plivo accepts custom headers that start with 'X-PH'
+                headers: {
+                  'X-PH-ROOMNAME': roomName,
+                  'X-PH-DIALOUT-NUMBER': phoneNumber,
+                },
+                auth: {
+                  username: tbConfig.sipUsername,
+                  password: tbConfig.sipPassword,
+                },
+                secure: true,
+              };
+              otInstance.dial_P(sessionId, token, tbConfig.sipUri, options)
+                .then((sipCallData) => {
+                  dialedNumbersInfo[phoneNumber] = {};
+                  dialedNumbersInfo[phoneNumber].sessionId = sipCallData.sessionId;
+                  dialedNumbersInfo[phoneNumber].connectionId = sipCallData.connectionId;
+                  aRes.send(sipCallData);
+                })
+                .catch((error) => {
+                  logger.log('postRoomDial error', error);
+                  aRes.status(400).send(new ErrorInfo(400, 'An error ocurred while forwarding SIP Call'));
+                });
+            });
         }
-        const sipCall = tbConfig.otInstance.dial_P(sessionId, token, tbConfig.sipUri, options)
-          .then((sipCallData) => {
-            dialedNumbersInfo[phoneNumber] = {}
-            dialedNumbersInfo[phoneNumber].sessionId = sipCallData.sessionId;
-            dialedNumbersInfo[phoneNumber].connectionId = sipCallData.connectionId;
-            aRes.send(sipCallData);
-          })
-          .catch((error) => {
-            logger.log('postRoomDial error', error)
-            aRes.status(400).send(new ErrorInfo(400, 'An error ocurred while forwarding SIP Call'));
-          });
-     });
+      })
+      .catch((err) => {
+        logger.log('postRoomDial => authentication error: ', err);
+        aRes.status(401).send(new ErrorInfo(100, 'Authentication Error'));
```

Comment:
@jeffswartz do you know what the difference is between the status code and the error code? 
Are the error codes defined somewhere?

Reply:
Also these error messages I just made up too

Reply:
Status code is the HTTP status code, while error codes are our own. I don't think there's a list. I usually used some variant of the error (so if I was returning a 403 I would use 40301, 40302 and so on to distinguish similar but different errors).
Also I think in this case you want a 403, not a 401. 401 means the user must provide some authentication (and it's usually used with basic authentication or digest authentication to show the prompt). 403 means the user did provide some authentication credentials, we just don't like them.

Reply:
Thanks explaining how these work.

With regard to the status code, having read several online sources I'm of the opinion that 401 is for two scenarios:
- missing authentication
- failed authentication

And I believe this case to be failed as the google library has thrown some error when verifying the token.
Whereas 403 I would say is for:
- Valid and successful authentication but without access rights

I use a 403 later on when the login token is verified but isn't from our hosted domain @tokbox.com (set in config).

The difference for me:
401, we still don't know who you are
403, we know who you are, and this isn't for you